### PR TITLE
Feature: Flutter NavigationObserver

### LINKED ISSUE #182: [sentry-flutter] Support for navigation Breadcrumbs
I would like to request support for navigation breadcrumbs.

The motivation is, that a developer can better recreate the user flow to track down the reported crash.

This can be implemented using Flutters [NavigatorObserver](https://api.flutter.dev/flutter/widgets/NavigatorObserver-class.html) or [RouteObserver](https://api.flutter.dev/flutter/widgets/RouteObserver-class.html).

A very simplisitic example of this feature could like this:
(Please note that I haven't confirmed if this actually compiles. It is just meant as an illustration of possible solution.)
```dart
import 'package:flutter/widgets.dart';
import 'package:sentry_flutter/sentry_flutter.dart';

Future<void> main() async {
  await SentryFlutter.init(
    (options) {
      options.dsn = 'https://example@sentry.io/add-your-dsn-here';
      // For better groupping, change the 'example' below with your own App's package.
      options.addInAppInclude('sentry_flutter_example');
    },
    () {
        runApp(MyApp())
    },
   );
  }
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      navigatorObservers: [
        SentryNavigatorObserver(),
      ],
      theme: ThemeData.dark().copyWith(scaffoldBackgroundColor: darkBlue),
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class SentryNavigatorObserver extends NavigatorObserver {
  void didPush(Route<dynamic> route, Route<dynamic> previousRoute) {
    addBreadcrumb(route.settings.name, "wasPushed");
  }

  void didPop(Route<dynamic> route, Route<dynamic> previousRoute) {
    addBreadcrumb(route.settings.name, "wasPopped");
  }

  void didRemove(Route<dynamic> route, Route<dynamic> previousRoute) {
    addBreadcrumb(route.settings.name, "wasRemoved");
  }

  void didReplace({ Route<dynamic> newRoute, Route<dynamic> oldRoute }) {
    addBreadcrumb(newRoute.settings.name, "didReplace");
  }

  void addBreadCrumb(String routeName, String type) {
    Sentry.addBreadcrumb(Breadcrumb(
      message: routeName,
      category: "navigation",
      type: type,
    ));
  }

  void didStartUserGesture(Route<dynamic> route, Route<dynamic> previousRoute) { 
        // no breadcrumb needed
  }

  void didStopUserGesture() {
    // no breadcrumb needed
  }
}
```

### PR DESCRIPTION
**Pull request by ueman**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
I've introduced `SentryNavigationObserver` to record navigation events as breadcrumbs.

## :bulb: Motivation and Context
This adds navigation breadcrumbs for crash reports.
It fixes https://github.com/getsentry/sentry-dart/issues/182

## :green_heart: How did you test it?
I've written some tests.
However I would need some guidance to make the tests more throughout.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps
I would like to request some guidance on what tests to write.

### CODE REVIEW

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigation_observer.dart]: the idea here would be using an `Integration` (default_integrations.dart) that takes a `Hub` and `SentryOptions`,
also adding a new `bool` to `SentryOptions` so we can turn it off if necessary.
Then adding the `Integration` at `_addDefaultIntegrations` (sentry_flutter.dart)

See how we do on Android:

https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/ActivityBreadcrumbsIntegration.java

doing this, the Integration will be registered automatically, no need of a single line of code on the client.

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigation_observer.dart]: ideally, we use the same values for `Breadcrumbs`, See on Android

https://github.com/getsentry/sentry-java/blob/main/sentry-android-core/src/main/java/io/sentry/android/core/ActivityBreadcrumbsIntegration.java#L63-L67

type= navigation
state= didPop, didReplace...
screen=name of the screen
category=ui.lifecycle
and adding more fields if necessary, so our UI will be ready to display nicely.

File: flutter/lib/src/navigation/navigation_breadcrumb.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/navigation_breadcrumb.dart]: I like the idea, but I think having a class for each type of breadcrumb is a bit overkill,
I'd be in to expose a nice Factory or named constructor in the Breadcrumb class to take parameters instead of setting values manually, See this on Java:

https://github.com/getsentry/sentry-java/blob/3fad970bef0675f63fcfe63cd2ff871b62743ed6/sentry/src/main/java/io/sentry/Breadcrumb.java#L46-L60

**Comment by marandaneto** [comment]
I've removed the WIP title and made it a Draft

**Comment by marandaneto** [comment]
thanks so much @ueman this is going in a nice direction :)
with a little bit of work, this can land on the next version, let me know if you need any help.

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: or using the `navigation` type: see https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/

```
{
  "type": "navigation",
  "category": "navigation",
  "timestamp": "2016-04-20T20:55:53.845Z",
  "data": {
    "from": "/login",
    "to": "/dashboard"
  }
}
```

File: flutter/lib/src/navigation/navigation_breadcrumb.dart
Comment (ueman) [reply flutter/lib/src/navigation/navigation_breadcrumb.dart]: `Breadcrumb` is a class of sentry-dart, so I can't add a constructor or factory with `RouteSettings` parameters without adding a Flutter dependency in sentry-dart. So this was my next best idea.

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (ueman) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: Thanks, will do.

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (ueman) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: A navigation observer must unfortunately be added manually as they can't be added automatically.
See https://api.flutter.dev/flutter/material/MaterialApp/navigatorObservers.html

```dart
MaterialApp(
   navigatorObservers: [
     SentryNavigatorObserver(),
   ],
   // other parameter ...
)
```

So I guess the `Integration` and the `SentryOptions` option are unnecessary?
I would however add this to the readme so users are aware of it.

Comment (ueman) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: I guess the `ActivityBreadcrumbsIntegration.java` is more equivalent to the [`WidgetsBindingObserver`](https://api.flutter.dev/flutter/widgets/WidgetsBindingObserver-class.html) (this would be a nice addition in another PR) than the `NavigationObserver`.

Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: > I guess the `ActivityBreadcrumbsIntegration.java` is more equivalent to the [`WidgetsBindingObserver`](https://api.flutter.dev/flutter/widgets/WidgetsBindingObserver-class.html) (this would be a nice addition in another PR) than the `NavigationObserver`.

true, I was just trying to give examples on how to use the `Integration` feature

File: flutter/lib/src/navigation/navigation_breadcrumb.dart
Comment (marandaneto) [reply flutter/lib/src/navigation/navigation_breadcrumb.dart]: yeah I do get that, but I'd not pass the whole `RouteSettings` but rather the values (as parameters), so if another Router FW is gonna be integrated, the facade is already available thru the `Breadcrumb` class.

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigation_observer.dart]: instead of using the `Sentry.currentHub` I'd rather use the `HubAdapter` which is basically redirecting to `Sentry.currentHub`, so we don't have a static class reference and still, unit testable.

Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigation_observer.dart]:
```suggestion
  factory SentryNavigatorObserver({Hub hub = HubAdapter()}) {
    return SentryNavigatorObserver._(hub);
  }
```

just an example.

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (ueman) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: Ah, I see. So what do you suggest on how to make the `SentryNavigationObserver` available to users?

Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: > A navigation observer must unfortunately be added manually as they can't be added automatically.
> See https://api.flutter.dev/flutter/material/MaterialApp/navigatorObservers.html
> 
> ```dart
> MaterialApp(
>    navigatorObservers: [
>      SentryNavigatorObserver(),
>    ],
>    // other parameter ...
> )
> ```
> 
> So I guess the `Integration` and the `SentryOptions` option are unnecessary?
> I would however add this to the readme so users are aware of it.

Oh I see, that makes sense, yeah if the user needs to add manually, then for sure `Integration` and `SentryOptions` are unnecessary, thanks for pointing this out.

Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: > Ah, I see. So what do you suggest on how to make the `SentryNavigationObserver` available to users?

reading the `NavigatorObserver` docs now I see that it'd not be possible to hook it up automatically unless we create our own `SentryMaterialApp` but I guess this is too much intrusive and if people need to create a new `SentryMaterialApp` anyway, I'd rather having less code and documenting they should do (what u have done on the README), so you are on the right path.

Comment (marandaneto) [reply flutter/lib/src/navigation/sentry_navigation_observer.dart]: thanks for clarifying btw, I didn't play enough with the Flutter UI yet.

File: flutter/test/navigation_observer_test.dart
Comment (marandaneto) [inline flutter/test/navigation_observer_test.dart]: some people are not fun of `Mocks` and I get that, but if you use the `MockHub` that already exists and `verify` the passed breadcrumb, you could `assert` the values having less boilerplate (all the `UnimplementedError` methods)

File: flutter/lib/src/navigation/sentry_navigation_observer.dart
Comment (marandaneto) [inline flutter/lib/src/navigation/sentry_navigation_observer.dart:44]: we are not using the null safety feature yet, wondering if this would cause problems, I'd rather just do null checks till the null-safety gets stable and we do a full migration

File: flutter/test/navigation_observer_test.dart
Comment (ueman) [reply flutter/test/navigation_observer_test.dart]: Done ✅

File: flutter/lib/src/navigation/navigation_breadcrumb.dart
Comment (ueman) [reply flutter/lib/src/navigation/navigation_breadcrumb.dart]: I would like to keep it as it is, because it is pretty Flutter specific and it makes the testing easier.

Also in future versions one could easily add more Flutter specific information. [`PageRoute`](https://api.flutter.dev/flutter/widgets/PageRoute-class.html) and co have much more information, which could potentially be useful. For example the name of the Route itself (eg. MaterialPageRoute, CupertinoPageRoute...) or if the navigation action covers the whole screen or if it is a dialog, bottom sheet or something else.

Comment (marandaneto) [reply flutter/lib/src/navigation/navigation_breadcrumb.dart]: so the code looks good, but we are coupling a class called `NavigationBreadcrumb` which is a pretty generic name for any navigation FW/3rd party lib (to the RouteObserver lib.)

I'd suggest moving all the code to the `SentryNavigatorObserver` class which is pretty much what is used for, don't get me wrong, the abstraction does make sense, but for an SDK that could be integrated with other 3rd parties not (IMO), as the `NavigationBreadcrumb` takes a `RouteSettings`.

so I see 2 alternatives, either we rename `NavigationBreadcrumb` to something more specific related to the `RouteObserver` or we move this logic to the `_addBreadcrumb` class, really your call, happy with both.

**Comment by marandaneto** [comment]
if you resolve the conflicts and mark it as ready to review, we could see if the CI is happy.

once more, thanks for contributing (L)

File: flutter/example/lib/main.dart
Comment (marandaneto) [inline flutter/example/lib/main.dart]: so not sure if there's a bug here or you did intentionally:

> Navigator operation requested with a context that does not include a Navigator.
> The context used to push or pop routes from the Navigator must be that of a widget that is a descendant of a Navigator widget.

**Comment by marandaneto** [comment]
I found a bug on the breadcrumbs parsing and the fix is landing here https://github.com/getsentry/sentry-dart/pull/201

File: flutter/example/lib/main.dart
Comment (ueman) [reply flutter/example/lib/main.dart]: Oh shoot, you're right. That was definitely a mistake on my side. I've fixed it.

Comment (marandaneto) [reply flutter/example/lib/main.dart]: cool, waiting for the CI #201 and you can merge main into your branch and GG :)

**Comment by marandaneto** [comment]
Screenshot 2020-11-29 at 10 53 26

**Review by marandaneto — APPROVED** [review]
LGTM

**Comment by marandaneto** [comment]
thanks a lot @ueman 
happy to merge when CI is happy.

**Pull request merged**
